### ISSUE TITLE
Hangfire.PostgreSql is making 25+ PostgreSql database connection 

### ISSUE THREAD

**Issue by k123009** [frankhommers/Hangfire.PostgreSql#140]
We are using PostgreSql as storage for Hangfire in our dot net core web api project. One thing which we have noticed is when we starts our Web Api application from IIS, it create 25+ database connections. Is this any thing which we should fix ? because making that huge number of connections is wastage of resource.

**Comment by frankhommers** [comment]
Isn't this just connection pooling?

https://www.npgsql.org/doc/connection-string-parameters.html#pooling

**Comment by willdean** [comment]
I am investigating this right now, and I think there may be a problem that Hangfire.PostgreSql is permanently holding onto a database connection for each Hangfire worker, at least at startup.

What I'm seeing is that, as startup, each Hangfire worker is sitting blocked in `PostgreSqlJobQueue.Dequeue_Transaction`, at this wait:

```
          WaitHandle.WaitAny(new WaitHandle[2]
          {
            cancellationToken.WaitHandle,
            (WaitHandle) PostgreSqlJobQueue.NewItemInQueueEvent
          }, this._options.QueuePollInterval);
          cancellationToken.ThrowIfCancellationRequested();
```

And they're holding onto a connection while they sit there.

For the purposes of investigation, I am playing games with npgsql, where I limit the max pool size to very small numbers (say five connections).  In this situation, the application cannot start because Hangfire (with the default 20 workers) consumes all available connections, and that then blocks calls like `RecurringJob.AddOrUpdate`, because they can't get a connection.  Eventually (15 seconds) npsql times-out and the application startup aborts.

This isn't a big problem to us (we don't need anything like 20 workers), but I don't imagine that the design intent was really to hold on to an incredibly expensive resource like a database connection for every waiting worker.  

I have a feeling that this somehow settles down once the system is running and there isn't a connection-per-worker consumed (we don't see 20 connections all the time in service), but I am still investigating this.

**Comment by willdean** [comment]
A minor update to my previous comment: it’s not that the connection usage settles down after startup, it was just that our servers have fewer processors than the dev machine I was testing on.  It does appear that there is a connection held open for each worker.

**Comment by davidroth** [comment]
Yes I confirm there is a design issue which leads to many open connections. I investigated this too because we did also notice some connection pooling issue in one of our applications during load.

As you have already mentioned, `PostgreSqlJobQueue.Dequeue_Transaction` spins in a loop in order to dequeue an item from the queue. The problem is, that while waiting (`this._options.QueuePollInterval`), the connection is still held in `Open` state. That has the unfortunate consequence, that each hangefire worker exclusively reserves a **physical** db connection. Because if this design issue, this increases pressure on the .net connection pool.

It gets worse:

The default hangfire worker count ist 25. That means that once the application is started, hangfire starts 25 workers, **each** of them opening a db connection and keeping each in `Open` state while sitting in the while loop (`PostgreSqlJobQueue.Dequeue_Transaction`). That means that hangfire already uses 25% of the typical postgres sql connection limit. Each additional application query cannot reuse the first 1-25 connections, because they are exclusively used by hangfire. This increases the risk of thread pool connection exhaustion.

A better design would `Open` the connection briefly before triggering the query and then immediately closing it afterwords. This would make the physical connection available for other consumers. 

That means that with the current design, i would encourage everybody to configure the worker count to a lower level.
